File: judge/src/tested/languages/java/templates/EvaluationResult.java
import java.util.ArrayList;
import java.util.List;

public class EvaluationResult {

    public final boolean result;
    public final String readableExpected;
    public final String readableActual;
    public final List<String> messages;

    private EvaluationResult(boolean result, String readableExpected, String readableActual, List<String> messages) {
        this.result = result;
        this.readableExpected = readableExpected;
        this.readableActual = readableActual;
        this.messages = messages;
    }

    public static Builder builder(boolean accepted) {
        return new Builder(accepted);
    }

    public static class Builder {

        private final boolean result;
        private String readableExpected = null;
        private String readableActual = null;
        private final List<String> messages = new ArrayList<>();

        private Builder(boolean result) {
            this.result = result;
        }

        public Builder withReadableExpected(String readableExpected) {
            this.readableExpected = readableExpected;
            return this;
        }

        public Builder withReadableActual(String readableActual) {
            this.readableActual = readableActual;
            return this;
        }

        public Builder withMessages(List<String> messages) {
            this.messages.addAll(messages);
            return this;
        }

        public Builder withMessage(String message) {
            this.messages.add(message);
            return this;
        }

        public EvaluationResult build() {
            return new EvaluationResult(result, readableExpected, readableActual, messages);
        }
    }
}

File: judge/src/tested/languages/kotlin/templates/EvaluationResult.kt
class EvaluationResult private constructor(
        val result: Boolean,
        val readableExpected: String?,
        val readableActual: String?,
        val messages: List<Message>
) {
    class Message(val description: String,
                  val format: String = "text",
                  val permission: String? = null)

    class Builder private constructor(
            private val result: Boolean,
            private var readableActual: String? = null,
            private var readableExpected: String? = null
    ) {
        private val messages: MutableList<Message> = ArrayList()

        fun withReadableActual(readableActual: String?): Builder {
            this.readableActual = readableActual
            return this
        }

        fun withReadableExpected(readableExpected: String?): Builder {
            this.readableExpected = readableExpected
            return this
        }

        fun withMessages(messages: List<Message>): Builder {
            this.messages += messages
            return this
        }

        fun withMessage(message: Message): Builder {
            this.messages += message
            return this
        }

        fun build(): EvaluationResult {
            return EvaluationResult(result, readableExpected,
                    readableActual, messages)
        }
    }
}
